sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Watch.java

```diff
@@ -725,25 +728,30 @@ public ProcessContinuation process(
         int numPending = tracker.addNewAsPending(res);
         if (numPending > 0) {
           LOG.info(
-              "{} - polling returned {} results, of which {} were new. The output is {}.",
+              "{} - current round of polling took {} ms and returned {} results, "
+                  + "of which {} were new. The output is {}.",
               c.element(),
+              new Duration(now, Instant.now()).getMillis(),
               res.getOutputs().size(),
               numPending,
               BoundedWindow.TIMESTAMP_MAX_VALUE.equals(res.getWatermark())
-                  ? "complete"
-                  : "incomplete");
+                  ? "final"
+                  : "not yet final");
         }
       }
-      while (tracker.hasPending()) {
+      int numEmitted = 0;
+      while (true) {
         c.updateWatermark(tracker.getWatermark());
-
-        TimestampedValue<OutputT> nextPending = tracker.tryClaimNextPending();
-        if (nextPending == null) {
-          return stop();
+        Map.Entry<HashCode, TimestampedValue<OutputT>> entry = tracker.getNextPending();
+        if (entry == null || !tracker.tryClaim(entry.getKey())) {
+          break;
         }
+        TimestampedValue<OutputT> nextPending = entry.getValue();
         c.outputWithTimestamp(
             KV.of(c.element(), nextPending.getValue()), nextPending.getTimestamp());
+        ++numEmitted;
       }
+      LOG.debug("{} - emitted {} new results.", c.element(), numEmitted);
```

Comment:
<!--new_thread; commit:b04f5ed31c0be1f231dd7c1613ea50d31616b127; resolved:0-->
This log might be a bit confusing. It says new results but do not reset numEmitted after this log.

Reply:
Made this log more comprehensive and added some clarifying variables.